### PR TITLE
Prefix names of inlined functions

### PR DESCRIPTION
**Pull request by alex-mckenna**
Identifiers which are assigned an inlined function now have a
prefix with the name of the function that was inlined. This makes
it clearer what the provinence of an assigment is in HDL.

### CODE REVIEW

**Comment by christiaanb** [comment]
```diff
diff --git a/clash-lib/src/Clash/Normalize/Transformations.hs b/clash-lib/src/Clash/Normalize/Transformations.hs
index 84cfcbe7..cadb0cf1 100644
--- a/clash-lib/src/Clash/Normalize/Transformations.hs
+++ b/clash-lib/src/Clash/Normalize/Transformations.hs
@@ -1722,9 +1722,9 @@ recToLetRec (TransformContext is0 []) e = do
       = and (zipWith (eqArg tcm) args args2)
     eqApp _ _ _ _ = False

-    eqArg _ v1 v2@(Var {})
+    eqArg _ v1 v2@(stripTicks -> Var {})
       = v1 == v2
-    eqArg tcm v1 v2@(collectArgs -> (Data _, args'))
+    eqArg tcm v1 v2@(collectArgs . stripTicks -> (Data _, args'))
       | let t1 = termType tcm v1
       , let t2 = termType tcm v2
       , t1 == t2                                                                                                           , t1 == t2  
```

Fixes `Feedback/Fib.hs`

**Comment by alex-mckenna** [comment]
Your ability to conjure fixes out of thin air never ceases to amaze :+1:

**Comment by christiaanb** [comment]
The name conflicts are a result in not being in normal form, I'm seeing:

```
LambdaDrop.core9185 after normalization:

λ(x1 :: Clash.Signal.Internal.Signal8214565720324203345
          "System"
          (GHC.Maybe.Maybe3674937295934324792
             Clash.Sized.Internal.BitVector.Bit8214565720324203376)) ->
letrec
  result13254 :: Clash.Sized.Internal.BitVector.Bit8214565720324203376
  = <prefixName>"core1_"
  letrec
    x6989586621679501791 :: Clash.Sized.Internal.BitVector.Bit8214565720324203376
    = case x1[LocalId] of
        GHC.Maybe.Just3891110078048108571
          (c$sel13252 :: Clash.Sized.Internal.BitVector.Bit8214565720324203376) ->
          c$sel13252[LocalId]
  in case x1[LocalId] of
       GHC.Maybe.Nothing3891110078048108568  ->
         Clash.Sized.Internal.BitVector.low
       GHC.Maybe.Just3891110078048108571
         (x6989586621679501795 :: Clash.Sized.Internal.BitVector.Bit8214565720324203376) ->
         x6989586621679501791[LocalId]
in result13254[LocalId]
```

i.e. a nested-letbinding. Haven't figured out why `flattenLet` is failing though.

**Comment by christiaanb** [comment]
Okay, it's not `flattenLet` that's failing to fire; it's `topLet` firing too early: it needs to fire after all lambda's and ticks:

```diff
diff --git a/clash-lib/src/Clash/Normalize/Transformations.hs b/clash-lib/src/Clash/Normalize/Transformations.hs
index 84cfcbe7..860ea32b 100644
--- a/clash-lib/src/Clash/Normalize/Transformations.hs
+++ b/clash-lib/src/Clash/Normalize/Transformations.hs
@@ -715,7 +715,7 @@ nonRepANF _ e = return e
 -- the body is a variable-reference.
 topLet :: HasCallStack => NormRewrite
 topLet (TransformContext is0 ctx) e
-  | all (\c -> isLambdaBodyCtx c || isTickCtx c) ctx && not (isLet e)
+  | all (\c -> isLambdaBodyCtx c || isTickCtx c) ctx && not (isLet e) && not (isTick e)
   = do
   untranslatable <- isUntranslatable False e
   if untranslatable
@@ -723,6 +723,9 @@ topLet (TransformContext is0 ctx) e
     else do tcm <- Lens.view tcCache
             argId <- mkTmBinderFor is0 tcm (mkUnsafeSystemName "result" 0) e
             changed (Letrec [(argId, e)] (Var argId))
+ where
+  isTick Tick {} = True
+  isTick _ = False

 topLet (TransformContext is0 ctx) e@(Letrec binds body)
   | all (\c -> isLambdaBodyCtx c || isTickCtx c) ctx
```

**Comment by christiaanb** [comment]
And finally, `splitNormalized` should take ticked `Letrec`s into account:

```diff
diff --git a/clash-lib/src/Clash/Netlist/Util.hs b/clash-lib/src/Clash/Netlist/Util.hs
index 7cadafed..a727bfbd 100644
--- a/clash-lib/src/Clash/Netlist/Util.hs
+++ b/clash-lib/src/Clash/Netlist/Util.hs
@@ -70,13 +70,13 @@ import           Clash.Core.Subst
    extendInScopeIdList, mkSubst, substTm)
 import           Clash.Core.Term
   (Alt, LetBinding, Pat (..), Term (..), TickInfo (..), NameMod (..),
-   collectArgsTicks)
+   collectArgsTicks, collectTicks)
 import           Clash.Core.TyCon
   (TyConName, TyConMap, tyConDataCons)
 import           Clash.Core.Type         (Type (..), TypeView (..),
                                           coreView1, splitTyConAppM, tyView, TyVar)
 import           Clash.Core.Util
-  (collectBndrs, stripTicks, substArgTys, termType, tyLitShow)
+  (collectBndrs, stripTicks, substArgTys, termType, tyLitShow, mkTicks)
 import           Clash.Core.Var
   (Id, Var (..), mkLocalId, modifyVarName, Attr')
 import           Clash.Core.VarEnv
@@ -129,9 +129,9 @@ splitNormalized
   -> Term
   -> (Either String ([Id],[LetBinding],Id))
 splitNormalized tcm expr = case collectBndrs expr of
-  (args,Letrec xes e)
+  (args,collectTicks -> (Letrec xes e,ticks))
     | (tmArgs,[]) <- partitionEithers args -> case stripTicks e of
-        Var v -> Right (tmArgs,xes,v)
+        Var v -> Right (tmArgs,map (second (`mkTicks` ticks)) xes,v)
         _     -> Left ($(curLoc) ++ "Not in normal form: res not simple var")
     | otherwise -> Left ($(curLoc) ++ "Not in normal form: tyArgs")
   _ ->
```

File: clash-lib/src/Clash/Netlist/Util.hs [hunk 1]
```diff
@@ -129,9 +129,9 @@ splitNormalized
   -> Term
   -> (Either String ([Id],[LetBinding],Id))
 splitNormalized tcm expr = case collectBndrs expr of
-  (args,Letrec xes e)
+  (args, collectTicks -> (Letrec xes e, ticks))
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Netlist/Util.hs]: My understanding is: ticks can wrap _any_ term, so any time you match on a term you should make sure that

* it either has no ticks
* or to call `collectTicks`

This seems like an impossible thing to do correctly, so ideally I'd like to see a solution where the compiler forces us to handle ticks. Could we think about this please? (Not for this PR though.)

File: clash-lib/src/Clash/Normalize.hs [hunk 2]
```diff
@@ -19,6 +19,9 @@
 
 module Clash.Normalize where
 
+import Debug.Trace -- TODO
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Normalize.hs]: TODO? :P

File: clash-lib/src/Clash/Normalize.hs [hunk 3]
```diff
@@ -323,6 +331,7 @@ flattenCallTree (CBranch (nm,(nm',sp,inl,tm)) used) = do
     _  -> do
       -- To have a cheap `appProp` transformation we need to
       -- deshadow, see also Note [AppProp no-shadow invariant]
+      -- TODO Tick inline?
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Normalize.hs]: What do you mean with "tick inline"?

File: clash-lib/src/Clash/Normalize/Transformations.hs [hunk 4]
```diff
@@ -377,18 +379,18 @@ inlineNonRep (TransformContext localScope _) e@(Case scrut altsTy alts)
...
+                      ," already inlined " ++ show limit ++ " times in:"
+                      , showPpr (varName cf)
+                      , "\nType of the subject is: " ++ showPpr scrutTy
+                      , "\nFunction " ++ showPpr (varName cf)
+                      , " will not reach a normal form, and compilation"
+                      , " might fail."
+                      , "\nRun with '-fclash-inline-limit=N' to increase"
+                      , " the inlining limit to N."
+                      ])
+              (return e)
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Normalize/Transformations.hs]: We should spend some time on building a quasiquoter.. (Again, not for this PR.)

File: clash-lib/src/Clash/Normalize/Transformations.hs [hunk 5]
```diff
@@ -399,9 +401,16 @@ inlineNonRep (TransformContext localScope _) e@(Case scrut altsTy alts)
         case (nonRepScrut, bodyMaybe) of
           (True,Just (_,_,_,scrutBody0)) -> do
             Monad.when noException (zoomExtra (addNewInline f cf))
+
+            traceM  $ "inlineNonRep: " <> (Text.unpack . nameOcc $ varName f)
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Normalize/Transformations.hs]: ^

File: clash-lib/src/Clash/Normalize/Transformations.hs [hunk 6]
```diff
@@ -968,8 +980,13 @@ inlineWorkFree (TransformContext localScope _) e@(collectArgsTicks -> (Var f,arg
             if isRecBndr
                then return e
                else do
-                 -- See Note [AppProp no-shadow invariant]
-                 changed (mkApps (mkTicks (deShadowTerm localScope body) ticks) args)
+                 traceM  $ "inlineWorkFree: " <> (Text.unpack . nameOcc $ varName f)
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Normalize/Transformations.hs]: ^

File: clash-lib/src/Clash/Normalize/Transformations.hs [hunk 7]
```diff
@@ -1024,9 +1041,19 @@ inlineSmall (TransformContext localScope _) e@(collectArgsTicks -> (Var f,args,t
           isRecBndr <- isRecursiveBndr f
           if not isRecBndr && inl /= NoInline && termSize body < sizeLimit
              then do
+               -- TODO This currently doesn't change clashing names when a
+               -- tick is added, causing tests like LambdaDrop to fail.
+
+               traceM  $ "inlineSmall: " <> (Text.unpack . nameOcc $ varName f)
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Normalize/Transformations.hs]: ^

File: clash-lib/src/Clash/Normalize/Util.hs [hunk 8]
```diff
@@ -432,3 +435,11 @@ removedTm
   -> Term
 removedTm =
   TyApp (Prim "Clash.Transformations.removedArg" (PrimInfo undefinedTy WorkNever))
+
+-- | A tick to prefix an inlined expression with it's original name.
+--
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Normalize/Util.hs]: An example would be nice here

File: clash-lib/src/Clash/Core/Evaluator.hs [hunk 9]
```diff
@@ -356,11 +357,18 @@ force :: Heap -> Stack -> Id -> Maybe State
 force (Heap gh g@(GPureHeap gbl) h ids is) k x' = case lookupVarEnv x' h of
     Nothing -> case lookupVarEnv x' gbl of
       Just e | isGlobalId x'
-        -> Just (Heap gh (GPureHeap (delVarEnv gbl x')) h ids is,GUpdate x':k,e)
+        -> let e' = tickExpr e
+            in do traceM $ "force: " <> (Text.unpack . nameOcc $ varName x')
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Core/Evaluator.hs]: ^

Comment (christiaanb) [reply clash-lib/src/Clash/Core/Evaluator.hs]: All of these traces will be removed for the final PR I assume?

Comment (martijnbastiaan) [reply clash-lib/src/Clash/Core/Evaluator.hs]: Ah right, I thought this was the final PR.

File: clash-lib/src/Clash/Normalize.hs (hunk 2 above)
Comment (alex-mckenna) [reply clash-lib/src/Clash/Normalize.hs]: I think we should have more random TODOs scattered in the code. After all, our work is never done :wink:

**Pull request merged**